### PR TITLE
DRA: structured "all" mode fix

### LINKED ISSUE #127554: DRA: Extra unexpected devices allocated when using 'allocationMode: All'
### What happened?

I was testing the ability to allocate both node-local resources (GPUs) along with a new network attached resource called an IMEX channel.

My setup is as follows:
* 8 nodes with 1 GPU each
* 2 pools of IMEX channels (with 10 available in each)
* IMEX channels from the first pool are part of "IMEX domain 1" and can be attached to nodes 0-3
* IMEX channels from the second pool are part of "IMEX domain 2" and can be attached to nodes 4-7

With this setup, I create two ResourceClaims, one ResourceClaim template, and 2 deployments. The two resource Claims are for two distinct IMEX channels, the ResourceClaimTemplate is for all GPUs on a given node, and the 2 deployments (of 4 replicas each) consume these claims / claimtemplates across each replica in order to simulate running 2 MPI jobs across two different IMEX domains.

Here are the specs:
```
---
apiVersion: v1
kind: Namespace
metadata:
  name: imex-test1
---
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceClaim
metadata:
  namespace: imex-test1
  name: shared-imex-channel0
spec:
  devices:
    requests:
    - name: channel
      deviceClassName: imex.nvidia.com
---
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceClaim
metadata:
  namespace: imex-test1
  name: shared-imex-channel1
spec:
  devices:
    requests:
    - name: channel
      deviceClassName: imex.nvidia.com
---
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceClaimTemplate
metadata:
  namespace: imex-test1
  name: all-node-gpus
spec:
  spec:
    devices:
      requests:
      - name: all-gpus
        deviceClassName: gpu.nvidia.com
        allocationMode: All
---
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: imex-test1
  name: pod0
  labels:
    app: imex-test1-pod0
spec:
  replicas: 4
  selector:
    matchLabels:
      app: pod0
  template:
    metadata:
      labels:
        app: pod0
    spec:
      containers:
      - name: ctr
        image: ubuntu:22.04
        command: ["bash", "-c"]
        args: ["trap 'exit 0' TERM; sleep 9999 & wait"]
        resources:
          claims:
          - name: gpus
          - name: imex-channel
      resourceClaims:
      - name: gpus
        resourceClaimTemplateName: all-node-gpus
      - name: imex-channel
        resourceClaimName: shared-imex-channel0
---
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: imex-test1
  name: pod1
  labels:
    app: imex-test1-pod1
spec:
  replicas: 4
  selector:
    matchLabels:
      app: pod1
  template:
    metadata:
      labels:
        app: pod1
    spec:
      containers:
      - name: ctr
        image: ubuntu:22.04
        command: ["bash", "-c"]
        args: ["trap 'exit 0' TERM; sleep 9999 & wait"]
        resources:
          claims:
          - name: gpus
          - name: imex-channel
      resourceClaims:
      - name: gpus
        resourceClaimTemplateName: all-node-gpus
      - name: imex-channel
        resourceClaimName: shared-imex-channel1
```

When I run this, I do not get the resource allocated to each pod in each deployment as expected (instead all pods remain pending forever).

However, If I change to explicitly requesting 1 GPU from a node instead of using `allocationMode: All`, things work as expected.

One thing to note is that the code linked below doesn't consider the CEL expression selector in the `gpu.nvidia.com` and `imex.nvidia.com` device classes when calculating `requestData.numDevices`. That might be factor in this somehow:
https://github.com/kubernetes/kubernetes/blob/52095a8b7b9b75d67a3882a21a6647e4f90ade48/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L176-L210

### What did you expect to happen?

My expectation was that we would see 4 pods from each deployment with the following set of resources:

**deployment 0, pod0**:
1 GPU from a node in the same IMEX domain as all other pods in deployment 0
IMEX channel 0 from the same IMEX domain as all other pods in deployment 0

**deployment 0, pod1**:
1 GPU from a node in the same IMEX domain as all other pods in deployment 0
IMEX channel 0 from the same IMEX domain as all other pods in deployment 0

**deployment 0, pod2**:
1 GPU from a node in the same IMEX domain as all other pods in deployment 0
IMEX channel 0 from the same IMEX domain as all other pods in deployment 0

**deployment 0, pod3**:
1 GPU from a node in the same IMEX domain as all other pods in deployment 0
IMEX channel 0 from the same IMEX domain as all other pods in deployment 0

**deployment 1, pod0**:
1 GPU from a node in the same IMEX domain as all other pods in deployment 1
IMEX channel 0 from the same IMEX domain as all other pods in deployment 1

**deployment 1, pod1**:
1 GPU from a node in the same IMEX domain as all other pods in deployment 1
IMEX channel 0 from the same IMEX domain as all other pods in deployment 1

**deployment 1, pod2**:
1 GPU from a node in the same IMEX domain as all other pods in deployment 1
IMEX channel 0 from the same IMEX domain as all other pods in deployment 1

**deployment 1, pod3**:
1 GPU from a node in the same IMEX domain as all other pods in deployment 1
IMEX channel 0 from the same IMEX domain as all other pods in deployment 1

### How can we reproduce it (as minimally and precisely as possible)?

Apply the specs listed above in a cluster with the following deviceClasses and DRA resources available...

Here are my (relevant) device classes:
```
---
- apiVersion: resource.k8s.io/v1alpha3
  kind: DeviceClass
  metadata:
    name: gpu.nvidia.com
  spec:
    selectors:
    - cel:
        expression: device.driver == 'gpu.nvidia.com' && device.attributes['gpu.nvidia.com'].type
          == 'gpu'
---
- apiVersion: resource.k8s.io/v1alpha3
  kind: DeviceClass
    name: imex.nvidia.com
  spec:
    selectors:
    - cel:
        expression: device.driver == 'gpu.nvidia.com' && device.attributes['gpu.nvidia.com'].type
          == 'imex-channel'
```

Here is the definition of one of my GPU nodes (the others look similar except for the node name):
```
- apiVersion: resource.k8s.io/v1alpha3
  kind: ResourceSlice
  metadata:
    name: k8s-dra-driver-cluster-worker-gpu.nvidia.com-dpwj8
  spec:
    devices:
    - basic:
        attributes:
          architecture:
            string: Ampere
          brand:
            string: Nvidia
          cudaComputeCapability:
            version: 8.0.0
          cudaDriverVersion:
            version: 12.6.0
          driverVersion:
            version: 560.35.3
          index:
            int: 0
          minor:
            int: 7
          productName:
            string: NVIDIA A100-SXM4-40GB
          type:
            string: gpu
          uuid:
            string: GPU-b1028956-cfa2-0990-bf4a-5da9abb51763
        capacity:
          memory: 40Gi
      name: gpu-0
    driver: gpu.nvidia.com
    nodeName: k8s-dra-driver-cluster-worker
    pool:
      generation: 0
      name: k8s-dra-driver-cluster-worker
      resourceSliceCount: 1
```

Here is the definition of my 2 IMEX channel resourceSlices:
```
---
apiVersion: v1
items:
- apiVersion: resource.k8s.io/v1alpha3
  kind: ResourceSlice
  metadata:
    name: imex-domain-0f884867-ba2f-4294-9155-b495ff367eea-1
  spec:
    devices:
    - basic:
        attributes:
          channel:
            int: 0
          type:
            string: imex-channel
      name: imex-channel-0
    - basic:
        attributes:
          channel:
            int: 1
          type:
            string: imex-channel
      name: imex-channel-1
    ...
    driver: gpu.nvidia.com
    nodeSelector:
      nodeSelectorTerms:
      - matchExpressions:
        - key: nvidia.com/gpu.clusteruuid
          operator: In
          values:
          - 0f884867-ba2f-4294-9155-b495ff367eea
        - key: nvidia.com/gpu.cliqueid
          operator: In
          values:
          - "1"
    pool:
      generation: 0
      name: imex-domain-0f884867-ba2f-4294-9155-b495ff367eea-1
      resourceSliceCount: 1
---
apiVersion: v1
items:
- apiVersion: resource.k8s.io/v1alpha3
  kind: ResourceSlice
  metadata:
    name: imex-domain-0f884867-ba2f-4294-9155-b495ff367eea-2
  spec:
    devices:
    - basic:
        attributes:
          channel:
            int: 0
          type:
            string: imex-channel
      name: imex-channel-0
    - basic:
        attributes:
          channel:
            int: 1
          type:
            string: imex-channel
      name: imex-channel-1
    ...
    driver: gpu.nvidia.com
    nodeSelector:
      nodeSelectorTerms:
      - matchExpressions:
        - key: nvidia.com/gpu.clusteruuid
          operator: In
          values:
          - 0f884867-ba2f-4294-9155-b495ff367eea
        - key: nvidia.com/gpu.cliqueid
          operator: In
          values:
          - "2"
    pool:
      generation: 0
      name: imex-domain-0f884867-ba2f-4294-9155-b495ff367eea-2
      resourceSliceCount: 1
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.31.1
Kustomize Version: v5.4.2
Server Version: v1.31.0
```

</details>

### Cloud provider

NONE

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Support for "allocationMode: all" was faulty and could have failed in different ways:
- Allocating devices not matching the selector in the device class.
- Not finding a solution when allocating more than one claim or request at a time.
- Reporting "stop allocation" as error instead of the normal "cannot allocate all claims".

#### Which issue(s) this PR fixes:
Fixes #127554

#### Does this PR introduce a user-facing change?
```release-note
DRA: fixed several issues related to "allocationMode: all"
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
```

/assign @klueska

### CODE REVIEW

**Comment by pohly** [comment]
/hold

I'll use this PR to extend code coverage of the allocator.

**Comment by pohly** [comment]
/retest

Unrelated flake ("k8s.io/kubernetes/pkg/kubelet/volumemanager: reconciler ").

File: staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go [hunk 1]
```diff
@@ -169,9 +169,13 @@ func (a *Allocator) Allocate(ctx context.Context, node *v1.Node) (finalResult []
 				return nil, fmt.Errorf("claim %s, request %s: could not retrieve device class %s: %w", klog.KObj(claim), request.Name, request.DeviceClassName, err)
 			}
 
+			// Start collecting information about the collect.
```
Comment (elezar) [inline staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:172]:
```suggestion
			// Start collecting information about the request.
```
?

File: staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go [hunk 2]
```diff
@@ -612,7 +621,6 @@ func (alloc *allocator) isSelectable(r requestIndices, slice *resourceapi.Resour
 		// Must be some future, unknown device type. We cannot select it.
 		// If we don't find anything else, then this will get reported
 		// in the final result, so remember that we skipped some device.
-		alloc.skippedUnknownDevice = true
```
Comment (klueska) [inline staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:615]: Is this comment still relevant?

Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:615]: No. Forgot to remove it. Fixed.

File: staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go (hunk 1 above)
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:172]: Yes, fixed.

**Comment by pohly** [comment]
/retest

Job definitions should be fixed now.

**Comment by klueska** [comment]
/lgtm
/approve

I can confirm this fixes the issue linked in the description

**Pull request merged**